ollama/_utils.py

```diff
@@ -0,0 +1,213 @@
+from __future__ import annotations
+
+from typing import Any, Callable, List, Mapping, Optional, Union, get_args, get_origin
+from ollama._types import Tool
+from collections.abc import Sequence, Set
+from typing import Dict, Set as TypeSet
+import sys
+
+# Type compatibility layer for Union and UnionType
+if sys.version_info >= (3, 10):
+  from types import UnionType
+
+  def is_union(tp: Any) -> bool:
+    return get_origin(tp) in (Union, UnionType)
+else:
+
+  def is_union(tp: Any) -> bool:
+    return get_origin(tp) is Union
+
+
+# Map both the type and the type reference to the same JSON type
+TYPE_MAP = {
+  # Basic types
+  int: 'integer',
+  'int': 'integer',
+  str: 'string',
+  'str': 'string',
+  float: 'number',
+  'float': 'number',
+  bool: 'boolean',
+  'bool': 'boolean',
+  type(None): 'null',
+  None: 'null',
+  'None': 'null',
+  # Collection types
+  list: 'array',
+  'list': 'array',
+  List: 'array',
+  'List': 'array',
+  Sequence: 'array',
+  'Sequence': 'array',
+  tuple: 'array',
+  'tuple': 'array',
+  set: 'array',
+  'set': 'array',
+  Set: 'array',
+  TypeSet: 'array',
+  'Set': 'array',
+  # Mapping types
+  dict: 'object',
+  'dict': 'object',
+  Dict: 'object',
+  'Dict': 'object',
+  Mapping: 'object',
+  'Mapping': 'object',
+  Any: 'string',
+  'Any': 'string',
+}
+
+
+def _get_json_type(python_type: Any) -> str | List[str]:
+  # Handle Optional types (Union[type, None] and type | None)
+  if is_union(python_type):
+    args = get_args(python_type)
+    # Filter out None/NoneType from union args
+    non_none_args = [arg for arg in args if arg not in (None, type(None))]
+    if non_none_args:
+      if len(non_none_args) == 1:
+        return _get_json_type(non_none_args[0])
+      # For multiple types (e.g., int | str | None), return array of types
+      return [_get_json_type(arg) for arg in non_none_args]
+    return 'null'
+
+  # Handle generic types (List[int], Dict[str, int], etc.)
+  if get_origin(python_type) is not None:
+    # Get the base type (List, Dict, etc.)
+    base_type = TYPE_MAP.get(get_origin(python_type), None)
```

Comment:
Given the structure of the JSON we need:
eg: 
```JSON
{
    "type": "function",
    "function": {
        "name": "add_numbers",
        "description": "Add two numbers together.",
        "parameters": {
            "type": "object",
            "required": [
                "x"
            ],
            "properties": {
                "x": {
                    "type": "integer",
                    "description": "The first number"
                },
                "y": {
                    "type": "integer",
                    "description": "The second number"
                }
            }
        },
        "return_type": "integer"
    }
}
```
The `parameters` key has a type of `JsonSchemaValue` defined in `_types.py` where pydantic doesn't do any serialization on the string values which should be mapped to valid JSON. So this means if it sees `typing.List` as the type in python it leaves it as is rather than mapping it to `array`.